### PR TITLE
configure: Support for libpng-config

### PR DESCRIPTION
**Pull request by HuidaeCho**
This PR is needed to use the OSGeo4W version of libpng, not the MSYS2 version.

Usage:
```bash
OSGEO4W_ROOT_MSYS=/c/OSGeo4W \
./configure \
--libpng-config=mswindows/osgeo4w/libpng-config
```
The included `libpng-config` is version-agnostic as long as `${OSGEO4W_ROOT_MSYS}/include/libpng${PNG_DLL_VERSION}/png.h` exists.

### CODE REVIEW

**Comment by HuidaeCho** [comment]
Maybe, ``mswindows/osgeo4w/libpng-config`` would be better...

**Comment by HuidaeCho** [comment]
Usage:
```bash
OSGEO4W_ROOT_MSYS=/c/OSGeo4W \
./configure \
--libpng-config=mswindows/osgeo4w/libpng-config
```
The included `libpng-config` is version-agnostic as long as `${OSGEO4W_ROOT_MSYS}/include/libpng${PNG_DLL_VERSION}/png.h` exists.

**Comment by HuidaeCho** [comment]
BTW, is `libpng-config` included in most Linux distributions and macOS? Safe to remove `--with-png-(includes|libs)`?

**Comment by HuidaeCho** [comment]
> BTW, is `libpng-config` included in most Linux distributions and macOS? Safe to remove `--with-png-(includes|libs)`?

Probably, not...
https://github.com/OSGeo/grass/actions/runs/3609604194/jobs/6082895084#step:11:188

**Comment by HuidaeCho** [comment]
`HAVE_PNG_H` is not defined for clang?
https://github.com/OSGeo/grass/actions/runs/3610010032/jobs/6083548001#step:10:160
```
ERROR: write_image: unknown file type: .png
```
from
https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/lib/pngdriver/write.c#L39-L44

**Comment by HuidaeCho** [comment]
`libpng-config` from [Ubuntu 20.04](https://packages.ubuntu.com/focal/libpng-dev) seems to be broken. It doesn't print an `-L` flag for `--ldflags`.
```bash
...
libs="-lpng16"
...
L_opts=
R_opts=
...
ldopts=""
...
    --ldflags)
        echo ${ldopts} ${L_opts} ${R_opts} {$libs}
```

**Review by nilason — DISMISSED** [review]
You need to change `HAVE_PNG_H` to `HAVE_PNG` in the files:

https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/lib/pngdriver/read.c#L33

https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/lib/pngdriver/write.c#L39

And then it works for me.

There is one difference to `PNGLIB` compared to before:

was `-L/opt/local/lib -lpng  -lz`
with this: `-L/opt/local/lib -lpng16`

there is a missing `-lz` flag. I'm not sure of the importance or consequence of this.

**Comment by nilason** [comment]
> `HAVE_PNG_H` is not defined for Ubuntu gcc? 

`HAVE_PNG_H` was defined before with: `LOC_CHECK_INCLUDES(png.h,PNG,$PNGINC)`, but not with this PR. `HAVE_PNG` serves the same purpose.

File: configure [hunk 1]
```diff
@@ -10127,158 +10230,6 @@ fi # $USE_TIFF
...
-
-
-ac_save_cppflags="$CPPFLAGS"
-CPPFLAGS="$PNGINC $CPPFLAGS"
-for ac_header in png.h
-do :
-  ac_fn_c_check_header_mongrel "$LINENO" "png.h" "ac_cv_header_png_h" "$ac_includes_default"
-if test "x$ac_cv_header_png_h" = xyes; then :
-  cat >>confdefs.h <<_ACEOF
-#define HAVE_PNG_H 1
```
Comment (HuidaeCho) [inline configure:10182]: This is it!

File: configure.ac [hunk 2]
```diff
@@ -873,6 +874,55 @@ AC_SUBST(HISTORYLIB)
...
+
+  AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <png.h>]], [[png_read_image(NULL, NULL);]])],[],[
+  AC_TRY_LINK([#include <png.h>],[png_read_image(NULL, NULL);],PNGLIB="$PNGLIB",[
+  AC_MSG_ERROR([*** Unable to locate libpng library.])
+  ])
+  ])
+  LIBS=${ac_save_libs}
+  CFLAGS=${ac_save_cflags}
+
+  AC_DEFINE(HAVE_PNG, 1, [define if PNG exists])
```
[the remark this answers is not in the document]
Comment (HuidaeCho) [reply configure.ac:919]: Tried to be consistent with the other defines.
https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/configure.ac#L525
https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/configure.ac#L533
https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/configure.ac#L618
https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/configure.ac#L761
https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/configure.ac#L905

and 18 more, but there is inconsistency...
https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/configure.ac#L1470
https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/configure.ac#L1475

[the remark this answers is not in the document]
Comment (HuidaeCho) [reply configure.ac]: Hmm.. <*.h> is different..

**Comment by HuidaeCho** [comment]
> there is a missing `-lz` flag. I'm not sure of the importance or consequence of this.

Not sure either, but if `-lz` is missing from the system `libpng-config`, maybe harmless.

**Comment by nilason** [comment]
> > there is a missing `-lz` flag. I'm not sure of the importance or consequence of this.
> 
> Not sure either, but if `-lz` is missing from the system `libpng-config`, maybe harmless.

Well, it *is* there. Using `libpng-config --static --ldflags` I get `-L/opt/local/lib -lpng16 -lz`.

it comes from:
`all_libs="-lpng16 -lz "`
originating from libpng source file `libpng-config.in`:
`all_libs="-lpng@PNGLIB_MAJOR@@PNGLIB_MINOR@ @LIBS@"`

$ZLIB="-lz" and $MATHLIB="" (on my system) was previously set as mandatory (and added) by autoconf with:
`LOC_CHECK_LIBS(png,png_read_image,PNG,$PNGLIB,PNGLIB,,$ZLIB $MATHLIB,)`
see:
https://github.com/OSGeo/grass/blob/e8f26dd6c3deb3c375c1dd973b60475d3cfba7ea/aclocal.m4#L145-L159

Maybe `libpng-config --static --ldflags` could be used instead of `libpng-config --ldflags`?


Re: HAVE_PNG vs HAVE_PNG_H. There is already a bunch of similar to the former one with e.g. HAVE_OGR, HAVE_GDAL, HAVE_PDAL. But no big issue, works either way.

File: configure.ac (hunk 2 above)
[the remark this answers is not in the document]
Comment (nilason) [reply configure.ac]: See also https://github.com/OSGeo/grass/blob/main/include/grass/config.h.in which is generated by autoheader. The descriptions in the style "Define to 1 if you have the <bzlib.h> header file." comes from autoconf.

**Comment by HuidaeCho** [comment]
> `libpng-config --static --ldflags`

That's interesting.. Only `--static --ldflags` includes `-lm -lz`. This is my output from Slackware:
```
-L/usr/lib64 -lpng16 -lm -lz -lm
```
`ldd /usr/lib64/libpng16.so` actually produces
```
        linux-vdso.so.1 (0x00007fff14bd0000)
        libz.so.1 => /lib64/libz.so.1 (0x00007f4860828000)
        libm.so.6 => /lib64/libm.so.6 (0x00007f48606e0000)
        libc.so.6 => /lib64/libc.so.6 (0x00007f4860501000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f48608ae000)
```
including `libz.so.1` and `libm.so.6`, so I guess `libpng-config --static --ldflags` would be safer although we're not really doing "static linking".

**Pull request merged**